src/main.rs

```diff
@@ -0,0 +1,238 @@
+#[poise::command(prefix_command)]
+pub async fn register(ctx: Context<'_>) -> Result<()> {
+    poise::builtins::register_application_commands_buttons(ctx).await?;
+    Ok(())
+}
+
+#[poise::command(prefix_command, slash_command, subcommands("search", "random"))]
+pub async fn wob(ctx: Context<'_>) -> Result<()> {
+    Ok(())
+}
+
+#[poise::command(prefix_command, slash_command)]
+pub async fn search(ctx: Context<'_>, terms: Vec<String>) -> Result<()> {
+    println!("{terms:?}");
+
+    let mut m = ctx
+        .send(CreateReply::default().content(format!("Searching for '{}' ...", terms.join(", "))))
+        .await?
+        .into_message()
+        .await?;
+
+    let entries = ctx.data().arcanum.search_entries(&terms).await?;
+
+    if entries.count == 0 {
+        m.edit(&ctx, EditMessage::new().content("No results found!"))
+            .await?;
+        return Ok(());
+    }
+
+    let mut cur_idx = 0;
+
+    let get_edit = |idx: usize| {
+        EditMessage::new()
+            .content("")
+            .embed(ctx.data().arcanum.embed_entry(&entries.results[idx]))
+            .components(vec![
+                CreateActionRow::Buttons(vec![
+                    CreateButton::new("back_10")
+                        .disabled(idx < 10)
+                        .style(ButtonStyle::Secondary)
+                        .label("Back 10"),
+                    CreateButton::new("back")
+                        .disabled(idx == 0)
+                        .style(ButtonStyle::Secondary)
+                        .label("Back"),
+                    CreateButton::new("info")
+                        .style(ButtonStyle::Secondary)
+                        .disabled(true)
+                        .label(format!("{}/{}", idx + 1, entries.results.len())),
+                    CreateButton::new("next")
+                        .disabled(idx + 1 >= entries.results.len())
+                        .style(ButtonStyle::Secondary)
+                        .label("Next"),
+                    CreateButton::new("next_10")
+                        .disabled(idx + 10 >= entries.results.len())
+                        .style(ButtonStyle::Secondary)
+                        .label("Next 10"),
+                ]),
+                CreateActionRow::Buttons(vec![
+                    CreateButton::new("done")
+                        .style(ButtonStyle::Success)
+                        .label("Done"),
+                    CreateButton::new("delete")
+                        .style(ButtonStyle::Danger)
+                        .label("Delete"),
+                ]),
+            ])
+    };
+
+    m.edit(&ctx, get_edit(cur_idx)).await?;
+
+    let inner_context: &serenity::Context = ctx.as_ref();
+
+    let mut interaction_stream = m
+        .await_component_interaction(&inner_context.shard)
+        .timeout(Duration::from_secs(180))
+        .stream();
+
+    while let Some(interaction) = interaction_stream.next().await {
```

Comment:
I agree (and thanks for the other comments as well!). Pagination also has an extra complication with the fact that the API call only returns at most 250 results at a time, so this needs more thought (maybe an async iterator?). I do agree it needs a good abstraction around it, because pagination might come up elsewhere.